### PR TITLE
Release/0.3.4

### PR DESCRIPTION
**Pull request by Nashatyrev**
- Update Noise to the latest spec changes (https://github.com/libp2p/specs/tree/master/noise)
- The Noise is now compatible with the Go implementation 
- To make the Noise compatible with Lighthouse Libp2p fork (until the Noise is updated there), add the compatibility flag `NoiseXXSecureChannel.rustInteroperability`
- Make pubsub `AbstractRouter.getMessageId` method overridable.
-  Make PubsubApiImpl class overridable for potential customizing message composition (e.g. remove 'from' field)

### CODE REVIEW

File: src/test/kotlin/io/libp2p/pubsub/PubsubApiTest.kt [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.concurrent.TimeUnit
+
+class PubsubApiTest {
+
+    @Test
+    fun testNoFromMessageField() {
+        class CustomApi(router: PubsubRouter) : PubsubApiImpl(router) {
+            inner class CustomPublisher(privKey: PrivKey, seqId: Long) : PublisherImpl(privKey, seqId) {
+                override fun createMessageToSign(data: ByteBuf, vararg topics: Topic): Rpc.Message {
+                    // not including 'from' field
```
Comment (zilm13) [inline src/test/kotlin/io/libp2p/pubsub/PubsubApiTest.kt:29]: why?

Comment (Nashatyrev) [reply src/test/kotlin/io/libp2p/pubsub/PubsubApiTest.kt:29]: `From` field is subject to be removed at some point in the spec

File: src/main/kotlin/io/libp2p/security/noise/NoiseXXSecureChannel.kt [hunk 2]
```diff
@@ -153,11 +158,13 @@ private class NoiseIoHandshake(
...
-            if (payloadLength < 2) throw SecureHandshakeError()
-            payloadLength = payload.sliceArray(0..1).toUShortBigEndian()
-            if (payload.size < payloadLength + 2) throw SecureHandshakeError()
-            payload = payload.sliceArray(2 until payloadLength + 2)
+        if (NoiseXXSecureChannel.rustInteroperability) {
+            if (payloadLength > 0) {
+                if (payloadLength < 2) throw SecureHandshakeError()
+                payloadLength = payload.sliceArray(0..1).toUShortBigEndian()
+                if (payload.size < payloadLength + 2) throw SecureHandshakeError()
+                payload = payload.sliceArray(2 until payloadLength + 2)
```
[the remark this answers is not in the document]
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/security/noise/NoiseXXSecureChannel.kt:166]: No, it was ok in previous Noise spec revision to pad payload with random bytes. So it is correct just to skip them

[the remark this answers is not in the document]
Comment (zilm13) [reply src/main/kotlin/io/libp2p/security/noise/NoiseXXSecureChannel.kt:166]: @Nashatyrev got it!

**Pull request merged**